x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/service/ServiceAccountToken.java

```diff
@@ -0,0 +1,171 @@
+        this.secret = secret;
+    }
+
+    public ServiceAccountId getAccountId() {
+        return accountId;
+    }
+
+    public String getTokenName() {
+        return tokenName;
+    }
+
+    public SecureString getSecret() {
+        return secret;
+    }
+
+    public String getQualifiedName() {
+        return getAccountId().asPrincipal() + "/" + tokenName;
+    }
+
+    public SecureString asBearerString() throws IOException {
+        try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+            out.writeBytes(PREFIX);
+            out.write(getQualifiedName().getBytes(StandardCharsets.UTF_8));
+            out.write(':');
+            out.write(secret.toString().getBytes(StandardCharsets.UTF_8));
+            final String base64 = Base64.getEncoder().withoutPadding().encodeToString(out.toByteArray());
+            return new SecureString(base64.toCharArray());
+        }
+    }
+
+    public static ServiceAccountToken fromBearerString(SecureString bearerString) throws IOException {
+        final byte[] bytes = CharArrays.toUtf8Bytes(bearerString.getChars());
+        logger.trace("parsing token bytes {}", MessageDigests.toHexString(bytes));
+        try (InputStream in = Base64.getDecoder().wrap(new ByteArrayInputStream(bytes))) {
+            final byte[] prefixBytes = in.readNBytes(4);
+            if (prefixBytes.length != 4 || false == Arrays.equals(prefixBytes, PREFIX)) {
+                logger.trace(() -> new ParameterizedMessage(
+                    "service account token expects the 4 leading bytes to be {}, got {}.",
+                    Arrays.toString(PREFIX), Arrays.toString(prefixBytes)));
+                return null;
+            }
+            final char[] content = CharArrays.utf8BytesToChars(in.readAllBytes());
+            final int i = UsernamePasswordToken.indexOfColon(content);
+            if (i < 0) {
+                logger.trace("failed to extract qualified service token name and secret, missing ':'");
+                return null;
+            }
+            return new ServiceAccountToken(new String(Arrays.copyOfRange(content, 0, i)),
+                new SecureString(Arrays.copyOfRange(content, i + 1, content.length)));
+        }
+    }
+
+    @Override
+    public void close() {
+        secret.close();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o)
+            return true;
+        if (o == null || getClass() != o.getClass())
+            return false;
+        ServiceAccountToken that = (ServiceAccountToken) o;
+        return accountId.equals(that.accountId) && tokenName.equals(that.tokenName) && secret.equals(that.secret);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(accountId, tokenName, secret);
+    }
+
+    public static ServiceAccountToken newToken(ServiceAccountId accountId, String tokenName) {
+        return new ServiceAccountToken(accountId, tokenName, UUIDs.randomBase64UUIDSecureString());
```

Comment:
Should we validate the tokenName here?

Reply:
It is validated by the caller. But your comment prompted me to do some more thinkings. I now moved the validation logic inside the constructor so it is always triggered. 

As a side effect, this means when we try parse a service account token from bearer string, an invalid token name will cause the parse to fail and continue down to authentication chain. Before this change, the behaviour was authentication would fail (because there would be no match for the qualified token name) and authentication chain would stop. I think this behaviour change is OK and actually more consistent. Because we also have validation logic inside `ServiceAccountId` which would behave the same as the new behaviour. In general, I think parsing failure should include anything that is more of a format rule and not business specific. An differentiating example is the service accout namespace, currently it only supports `elastic`. But it is a business rule, not a formatting requirement.